File: modules/expo-bluesky-video-compress/android/src/main/java/expo/modules/blueskyvideocompress/ExpoBlueskyVideoCompressModule.kt
package expo.modules.blueskyvideocompress

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.util.concurrent.ConcurrentHashMap

class ExpoBlueskyVideoCompressModule : Module() {
  private val activeCompressors = ConcurrentHashMap<Int, VideoCompressor>()

  override fun definition() = ModuleDefinition {
    Name("ExpoBlueskyVideoCompress")

    Events("onProgress")

    AsyncFunction("probe") { uri: String ->
      val context = appContext.reactContext
        ?: throw Error("React context is null")
      return@AsyncFunction VideoProber.probe(context, uri)
    }

    AsyncFunction("compress") { uri: String, options: Map<String, Any?> ->
      val context = appContext.reactContext
        ?: throw Error("React context is null")
      val targetBitrate = (options["targetBitrate"] as? Number)?.toInt() ?: 0
      val maxSize = (options["maxSize"] as? Number)?.toInt() ?: 1920
      val codecPref = (options["codec"] as? String) ?: "auto"
      val frameRateCap = ((options["frameRateCap"] as? Number)?.toInt() ?: 30).coerceAtLeast(1)
      val jobId = (options["jobId"] as? Number)?.toInt() ?: 0

      val compressor = VideoCompressor(
        context = context,
        uriString = uri,
        targetBitrate = targetBitrate,
        maxSize = maxSize,
        codecPref = codecPref,
        frameRateCap = frameRateCap,
        jobId = jobId,
        onProgress = { id, progress ->
          sendEvent("onProgress", mapOf(
            "id" to id,
            "progress" to progress
          ))
        }
      )

      activeCompressors[jobId] = compressor

      try {
        val result = compressor.compress()
        activeCompressors.remove(jobId)
        return@AsyncFunction result
      } catch (e: Exception) {
        activeCompressors.remove(jobId)
        throw e
      }
    }

    Function("cancel") { jobId: Int ->
      activeCompressors.remove(jobId)?.cancel()
    }
  }
}

File: modules/expo-bluesky-video-compress/ios/ExpoBlueskyVideoCompressModule.swift
import AVFoundation
import ExpoModulesCore

public class ExpoBlueskyVideoCompressModule: Module {
  private var activeCompressors: [Int: VideoCompressor] = [:]
  private let activeCompressorsLock = NSLock()

  public func definition() -> ModuleDefinition {
    Name("ExpoBlueskyVideoCompress")

    Events("onProgress")

    AsyncFunction("probe") { (uri: String) -> [String: Any] in
      let url = URL(string: uri) ?? URL(fileURLWithPath: uri)
      return try await VideoProber.probe(url: url)
    }

    AsyncFunction("compress") { (uri: String, options: [String: Any]) -> [String: Any] in
      let url = URL(string: uri) ?? URL(fileURLWithPath: uri)
      let targetBitrate = options["targetBitrate"] as? Int ?? 0
      let maxSize = options["maxSize"] as? Int ?? 1920
      let codecPref = options["codec"] as? String ?? "auto"
      let frameRateCap = max(1, options["frameRateCap"] as? Int ?? 30)
      let jobId = options["jobId"] as? Int ?? 0

      let compressor = VideoCompressor(
        url: url,
        targetBitrate: targetBitrate,
        maxSize: maxSize,
        codecPref: codecPref,
        frameRateCap: frameRateCap,
        jobId: jobId,
        onProgress: { [weak self] id, progress in
          self?.sendEvent("onProgress", [
            "id": id,
            "progress": progress,
          ])
        }
      )

      self.setCompressor(jobId, compressor)

      do {
        let result = try await compressor.compress()
        self.setCompressor(jobId, nil)
        return result
      } catch {
        self.setCompressor(jobId, nil)
        throw error
      }
    }

    Function("cancel") { (jobId: Int) in
      self.cancelCompressor(jobId)
    }
  }

  private func setCompressor(_ jobId: Int, _ compressor: VideoCompressor?) {
    activeCompressorsLock.lock()
    defer { activeCompressorsLock.unlock() }
    if let compressor = compressor {
      activeCompressors[jobId] = compressor
    } else {
      activeCompressors.removeValue(forKey: jobId)
    }
  }

  private func cancelCompressor(_ jobId: Int) {
    activeCompressorsLock.lock()
    let compressor = activeCompressors.removeValue(forKey: jobId)
    activeCompressorsLock.unlock()
    compressor?.cancel()
  }
}
